### PR TITLE
GMP: New version 6.3.0

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

File: G/GMP/GMP@6.3.0/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+version = v"6.3.0"
+
+include("../common.jl")
+
+# Build the tarballs!
+build_tarballs(ARGS, configure(version)...;
+               clang_use_lld=false, julia_compat="1.6", preferred_gcc_version=v"6")
```
Comment (giordano) [inline G/GMP/GMP@6.3.0/build_tarballs.jl:7]: `clang_use_lld` doesn't work yet here, in one or two days probably 🙂

Comment (giordano) [reply G/GMP/GMP@6.3.0/build_tarballs.jl:7]: It should work now!

Comment (giordano) [reply G/GMP/GMP@6.3.0/build_tarballs.jl:7]: Yes, it does! Yay!

**Comment by eschnett** [comment]
@maleadt @ViralBShah There is a patch for GMP 6.2.1 that handles "overflow" (out of memory). The respective code changed in GMP 6.3.0. The older patches don't apply, and I don't know whether the patch should be updated or whether it is now superfluous.

**Comment by eschnett** [comment]
Since Julia Base calls `__gmp_set_alloc_overflow_function` I decided to undo the changes to GMP that now handle mpz overflow via an error code in favour of the mechanism introduced in our patches. The main reasons being (a) it will work for Julia right away, and (b) our patches handle more cases.

**Comment by maleadt** [comment]
I did add tests for the overflow cases I encountered to https://github.com/JuliaLang/julia/pull/51243, so as long as Julia tests pass it should be OK to go with upstream.

**Comment by giordano** [comment]
Msan build [is failing](https://buildkite.com/julialang/yggdrasil/builds/7492#018ccaed-2f81-4dc8-b798-2fe8ff809e0b/619-4270) with
```
[16:11:08] configure:11251: checking C++ compiler c++  -g -O2
[16:11:08] Test compile:
[16:11:08] configure:11265: c++  -g -O2 conftest.cc >&5
[16:11:08] /opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: warning: libc.musl-x86_64.so.1, needed by /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/../../../../x86_64-linux-musl/lib/../lib64/libstdc++.so, not found (try using -rpath or -rpath-link)
[16:11:08] /opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: a.out: hidden symbol `fstat' in /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib64/libc_nonshared.a(fstat.oS) is referenced by DSO
[16:11:08] /opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: final link failed: Bad value
[16:11:08] clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
[16:11:08] configure:11268: $? = 1
[16:11:08] failed program was:
[16:11:08] 
[16:11:08] int main (void) { return 0; }
[16:11:08] configure:11380: result: no
[16:11:08] configure:11396: error: C++ compiler not available, see config.log for details
```

**Comment by eschnett** [comment]
Compiling a simple C++ program
```C++
int main (void) { return 0; }
```
 works. Linking it fails:
```
sandbox:${WORKSPACE}/srcdir/gmp-6.3.0 # c++  --verbose -O2 -pedantic -fomit-frame-pointer -m64 conftest.o
clang version 13.0.1 (/home/mose/.julia/dev/BinaryBuilderBase/deps/downloads/clones/llvm-project.git-1df819a03ecf6890e3787b27bfd4f160aeeeeacd50a98d003be8b0893f11a9be 75e33f71c2dae584b13a7d1186ae0a038ba98838)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/x86_64-linux-musl/bin
Found candidate GCC installation: /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0
Selected GCC installation: /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/opt/bin/x86_64-linux-gnu-libgfortran3-cxx11-sanitize+memory/ld.x86_64-linux-gnu" --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root -z now -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64/crt1.o /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64/crti.o /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/crtbegin.o -L/opt/x86_64-linux-gnu/lib/gcc/opt/x86_64-linux-gnu/lib/gcc -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib -L/opt/x86_64-linux-gnu/lib/gcc/x86_64-linux-gnu/6.1.0 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -L/opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0 -L/opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/../../../../x86_64-linux-musl/lib/../lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib/../lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64 -L/opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/../../../../x86_64-linux-musl/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib --whole-archive /opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan-x86_64.a --no-whole-archive --whole-archive /opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan_cxx-x86_64.a --no-whole-archive --export-dynamic conftest.o -lstdc++ -lm --no-as-needed -lpthread -lrt -lm -ldl -lgcc_s -lgcc -lc -lgcc_s -lgcc /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/crtend.o /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64/crtn.o
/opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: cannot find /opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan-x86_64.a: No such file or directory
/opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: cannot find /opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan_cxx-x86_64.a: No such file or directory
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
```

It is weird that the linker is trying to pull in files (`crtbegin.o`, `libclang_rt.msan-x86_64.a` etc.) from the directory `/opt/x86_64-linux-musl`. I would expect it to use the files to be in `/opt/x86_64-linux-gnu` instead.

**Comment by eschnett** [comment]
Is [this](https://github.com/JuliaPackaging/Yggdrasil/blob/b7f5e3c48f292078bbed4c9fdad071da7875c0bc/S/SuiteSparse/common.jl#L45) related?

**Comment by giordano** [comment]
> Is [this](https://github.com/JuliaPackaging/Yggdrasil/blob/b7f5e3c48f292078bbed4c9fdad071da7875c0bc/S/SuiteSparse/common.jl#L45) related?

I missed that, but I don't think it's related, should have just used llvm 13 until we can build compiler-rt for llvm 16 (#7877)

**Comment by giordano** [comment]
```console
sandbox:${WORKSPACE}/srcdir/gmp-6.3.0 # echo 'int main (void) { return 0; }' | SUPER_VERBOSE=1 c++ -x c++ - -g -O2
ccache /opt/x86_64-linux-musl/bin/clang++ -target x86_64-linux-gnu --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root -D_GLIBCXX_USE_CXX11_ABI=1 -stdlib=libstdc++ -march=x86-64 -mtune=generic -fsanitize=memory -isystem /opt/x86_64-linux-gnu/x86_64-linux-gnu/include/c++/6.1.0 -isystem /opt/x86_64-linux-gnu/x86_64-linux-gnu/include/c++/6.1.0/x86_64-linux-gnu -isystem /opt/x86_64-linux-gnu/x86_64-linux-gnu/include/c++/6.1.0/backward -isystem /opt/x86_64-linux-gnu/x86_64-linux-gnu/include -isystem /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/include -x c++ - -g -O2 -rtlib=libgcc -L/opt/x86_64-linux-gnu/lib/gcc/opt/x86_64-linux-gnu/lib/gcc -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib -L/opt/x86_64-linux-gnu/lib/gcc/x86_64-linux-gnu/6.1.0 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -fuse-ld=x86_64-linux-gnu -fsanitize=memory
/opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root -z now -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64/crt1.o /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64/crti.o /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/crtbegin.o -L/opt/x86_64-linux-gnu/lib/gcc/opt/x86_64-linux-gnu/lib/gcc -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib -L/opt/x86_64-linux-gnu/lib/gcc/x86_64-linux-gnu/6.1.0 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -L/opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0 -L/opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/../../../../x86_64-linux-musl/lib/../lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib/../lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64 -L/opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/../../../../x86_64-linux-musl/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib --whole-archive /opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan-x86_64.a --no-whole-archive --dynamic-list=/opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan-x86_64.a.syms --whole-archive /opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan_cxx-x86_64.a --no-whole-archive --dynamic-list=/opt/x86_64-linux-musl/lib/clang/13.0.1/lib/linux/libclang_rt.msan_cxx-x86_64.a.syms /tmp/--fa2059.o -lstdc++ -lm --no-as-needed -lpthread -lrt -lm -ldl -lgcc_s -lgcc -lc -lgcc_s -lgcc /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/crtend.o /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib/../lib64/crtn.o
/opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: warning: libc.musl-x86_64.so.1, needed by /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/6.1.0/../../../../x86_64-linux-musl/lib/../lib64/libstdc++.so, not found (try using -rpath or -rpath-link)
/opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: a.out: hidden symbol `fstat' in /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib64/libc_nonshared.a(fstat.oS) is referenced by DSO
/opt/x86_64-linux-gnu/bin/x86_64-linux-gnu-ld: final link failed: Bad value
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
```
If I understand it correctly, the linker invocation looks extremely wrong, it's mixing up glibc and musl directories and object files.

**Comment by giordano** [comment]
@gbaraldi this was broken around https://github.com/JuliaPackaging/BinaryBuilderBase.jl/commit/65d60010f45cbafa66c9809c9532759bf62c970b (https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/324), but honestly I can't quite tell why.  Note that on that specific commit the C compiler is also broken, this was then fixed by https://github.com/JuliaPackaging/BinaryBuilderBase.jl/commit/e76628b13c5a0d610281695197df0dfb2a3c3064 (https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/332), here only the C++ compiler is broken, so it's a bit complicated to figure out _what_ exactly caused the breakage.

**Comment by giordano** [comment]
@gbaraldi I can confirm restoring `--gcc-toolchain=/opt/x86_64-linux-gnu` fixes the issue: https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/324#discussion_r1468471541

**Comment by giordano** [comment]
@eschnett good to go now?

**Comment by eschnett** [comment]
@giordano Good to go.
